sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,346 @@
+            EventHubProducer = eventHubProducer ?? connection.CreateTransportProducer(null, retryPolicy);
+
+            ExpirationTimer = new Timer(PerformExpiration(),
+                                        null,
+                                        performExpirationPeriod.Value,
+                                        performExpirationPeriod.Value);
+        }
+
+        /// <summary>
+        ///   Retrieves a <see cref="PooledProducer" /> for the requested partition,
+        ///   creating one if needed or extending the expiration for an existing instance.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The unique identifier of a partition associated with the Event Hub.</param>
+        /// <param name="removeAfterDuration">The period of inactivity after which a <see cref="PoolItem" /> will become eligible for eviction. Overrides <see cref="PoolItem.DefaultRemoveAfterDuration" />.</param>
+        ///
+        /// <returns>A <see cref="PooledProducer" /> mapping to the partition id passed in as parameter.</returns>
+        ///
+        /// <remarks>
+        ///   There is a slight probability that the returned producer may be closed at any time
+        ///   after it is returned and the caller may want to handle that scenario.
+        /// </remarks>
+        ///
+        public virtual PooledProducer GetPooledProducer(string partitionId,
+                                                        TimeSpan? removeAfterDuration = default)
+        {
+            if (string.IsNullOrEmpty(partitionId))
+            {
+                return new PooledProducer(EventHubProducer);
+            }
+
+            var identifier = Guid.NewGuid().ToString();
+
+            var item = Pool.GetOrAdd(partitionId, id => new PoolItem(partitionId, Connection.CreateTransportProducer(id, RetryPolicy), removeAfterDuration));
+
+            // A race condition at this point may end with CloseAsync called on
+            // the returned PoolItem if it had expired. The probability is very low and
+            // possible exceptions should be handled by the invoking methods.
+
+            if (item.PartitionProducer.IsClosed == true || !item.ActiveInstances.TryAdd(identifier, 0))
+            {
+                identifier = Guid.NewGuid().ToString();
+                item = Pool.GetOrAdd(partitionId, id => new PoolItem(partitionId, Connection.CreateTransportProducer(id, RetryPolicy), removeAfterDuration));
+                item.ActiveInstances.TryAdd(identifier, 0);
+            }
+
+            item.ExtendRemoveAfter(removeAfterDuration);
+
+            return new PooledProducer(item.PartitionProducer, cleanUp: producer =>
+            {
+                Argument.AssertNotNull(producer, nameof(producer));
+
+                if (Pool.TryGetValue(partitionId, out PoolItem pooledItem))
+                {
+                    pooledItem.ExtendRemoveAfter(removeAfterDuration);
+                }
```

Comment:
I'm reasonably sure that we had a reason.   I won't lie to you, I can't remember why off the top of my head.   @albertodenatale do you remember?

Reply:
It's a fair point and it is not your memory at all @jsquire.

We discussed if we should wrap it in an if statement but not if we should remove it altogether. We convened that once an item has been removed from the pool (and not closed because in use), the last user would be in charge of closing it. Hence we would only extend its expiration if it is in the ```Pool```.

I might be missing completely the point here, apologies for that, but I believe the motivation of keeping the extension was to reduce the chances of a race to close the producer between the ```PooledProducer.CleanUp``` function and the ```PerformExpiration``` method.

By extending it at the start of the function, we are confident enough that did the ```PerformExpiration``` started it would not try to remove the ```PooledProducer```.

Not sure if a "_race to close the producer_" might be considered benign?

Reply:
Thank you....  That still makes sense to me.  😄